http_backend.go

```diff
@@ -170,7 +171,12 @@ func (h *httpBackend) Cache(request *http.Request, bodySize int, checkHeadersFun
 func (h *httpBackend) Do(request *http.Request, bodySize int, checkHeadersFunc checkHeadersFunc) (*Response, error) {
 	r := h.GetMatchingRule(request.URL.Host)
 	if r != nil {
-		r.waitChan <- true
+		select {
+		case r.waitChan <- true:
+		case <-request.Context().Done():
+			return nil, errors.New("context canceled; early return")
```

Comment:
I believe earlier version of this PR returned `ctx.Err()` instead. Why did you decide to return a new error instead?

Reply:
Yes, that was the case. Usually i would use wrap from `github.com/pkg/errors` and just add to the error but the library is not in use here.
My first try at writing the code I was asserting against the time that it took to run all the calls and check if they return asap after the cancel; worked fine locally but fail on CI and yea, check like that is really unreliable.
So my second attempt its based on the error message; to check if the error was the normal error from the request or if it was the early return; I had to make the early return message different from `ctx.Err()` cause that is what we get from the http client that tries to run with a cancelled request.

Reply:
Sorry for the late response. I don't think tests are good enough reason to return different errors (depending on where it happened) when context is cancelled. Are you aware of any other library that does *not* return `ctx.Err()` unchanged?

Reply:
>  My first try at writing the code I was asserting against the time that it took to run all the calls and check if they return asap after the cancel; worked fine locally but fail on CI and yea, check like that is really unreliable.

I suppose you could make a test like this:
- Make the test HTTP handler sleep before returning the response, maybe a second or more.
- Set low rate limit, say, 1 req/second.
- Spawn lots of goroutines. Doesn't have to be too much, though, maybe 20-30 or so is enough.
- Cancel context.
- Assert that cancellation does happen in 5 seconds. This value seems high enough so the test would be reliable, but low enough so the test won't accidentally pass without the fix.

What do you think?